File: ThreadingSimulator/ThreadingSimulator/Models/SemaphorBlockModel.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using ThreadingSimulator.Enums;

namespace ThreadingSimulator.Models
{
    [XmlInclude(typeof(BlockModel))]
    [XmlInclude(typeof(LoopBlockModel))]
    [XmlInclude(typeof(SemaphorBlockModel))]
    [XmlInclude(typeof(VariableBlockModel))]
    public class SemaphorBlockModel : BlockModel
    {
        public string SemaphorName { get; set; }
        public SemaphorOperationType SemaphorOperation { get; set; }

        [XmlIgnore]
        [JsonIgnore]
        public override bool IsValid
        {
            get
            {
                return !String.IsNullOrEmpty(SemaphorName);
            }
        }

        public SemaphorBlockModel() : base(BlockType.SEMAPHOR)
        {
        }
    }
}

File: GUI/src/app/Models/SemaphorBlockModel.ts
import { BlockModel } from "./BlockModel";
import { SemaphorOperationType } from "../Enums/SemaphorOperationType";
import { isNullOrUndefined } from "util";
import { BlockType } from "../Enums/BlockType";

export class SemaphorBlockModel extends BlockModel
{
    public SemaphorName: string;
    public SemaphorOperation: SemaphorOperationType = SemaphorOperationType.ENTER;

    public IsValid(): boolean
    {
        return !isNullOrUndefined(this.SemaphorName);
    }
   
    public Clone(): SemaphorBlockModel
    {
        return SemaphorBlockModel.Deserialize(this);
    }

    public static Deserialize(obj: any): SemaphorBlockModel
    {
        let trueObj: SemaphorBlockModel = new SemaphorBlockModel();

        trueObj.Type = obj.Type;
        trueObj.SemaphorName = obj.SemaphorName;
        trueObj.SemaphorOperation = obj.SemaphorOperation;

        return trueObj;
    }
	
    constructor()
    {
        super(BlockType.SEMAPHOR);
    }
}